Sources/swiftML/LinearRegression.swift

```diff
@@ -0,0 +1,124 @@
+    /// The number of iterations for gradient descent.
+    var iterationCount: Int
+    /// The linear rate gradient descent.
+    var learningRate: Float
+    /// Whether to calculate the intercept for this model.
+    var fitIntercept: Bool
+    /// The weights of the model.
+    public var weights: Tensor<Float>
+
+    /// Creates a linear regression model.
+    ///
+    /// - Parameters:
+    ///   - gradientDescent: The gradient descent or singular vector decomposition method to be
+    ///     used for learning. The default is `false`.
+    ///   - iterationCount: The number of iterations for gradient descent. The default is `1000`.
+    ///   - learningRate: The learning rate for gradient descent. The default is `0.001`.
+    ///   - fitIntercept: Whether to calculate the intercept for this model. If `false`, no
+    ///     intercept will be used in calculations. The default is `true`.
+    public init(
+        gradientDescent: Bool = false,
+        iterationCount: Int = 1000,
+        learningRate: Float = 0.001,
+        fitIntercept: Bool = true
+    ) {
+        precondition(iterationCount > 0, "Iteration count must be positive.")
+        precondition(learningRate >= 0, "Learning rate must be non-negative.")
+
+        self.gradientDescent = gradientDescent
+        self.iterationCount = iterationCount
+        self.learningRate = learningRate
+        self.weights = Tensor<Float>(0)
+        self.fitIntercept = fitIntercept
+    }
+
+    /// Initialize weights between `[-1/N, 1/N]`.
+    /// - Parameter featuresCount: The number of features in training data.
+    internal func initializeWeights(featuresCount: Int) {
+        // Randomly initialize weights.
+        var w = Tensor<Float>(zeros: [featuresCount, 1])
+        let limit: Float = 1 / sqrt(Float(featuresCount))
+        for i in 0..<featuresCount {
+            w[i] = Tensor<Float>([Float.random(in: -limit...limit)])
+        }
+        self.weights = w
+    }
+    
+    /// Fit a linear model.
+    ///
+    /// - Parameters:
+    ///   - data: Training data with shape `[sample count, feature count]`.
+    ///   - labels: Target value with shape `[sample count, 1]`.
+    public func fit(data: Tensor<Float>, labels: Tensor<Float>) {
+        precondition(data.shape[0] > 0, "Data must have a positive sample count.")
+        precondition(data.shape[1] >= 1, "Data must have feature count greater than one.")
+        precondition(labels.shape[0] > 0, "Labels must have a positive sample count.")
+        precondition(labels.shape[1] == 1, "Labels must have single target feature.")
+        precondition(data.shape[0] == labels.shape[0],
+                     "Data and labels must have the same sample count.")
+
+        var data: Tensor<Float> = data
+        if self.fitIntercept {
+            let ones = Tensor<Float>(ones: [data.shape[0], 1])
+            data = ones.concatenated(with: data, alongAxis: -1)
+        }
+        if self.gradientDescent == false {       
```

Comment:
It seems that `class LinearRegression` currently has two different "strategies":
1. Gradient descent: which uses `fitIntercept`, `weights`, `learningRate`, `iterations`.
2. Singular value decomposition: which uses `fitIntercept`, `weights`.

Right now, the strategy is controlled by a flag `var gradientDescent: Bool`, but this is not particularly safe. `gradientDescent` may be `false`, but `learningRate` and `iterations` can still be accessed.

---

To make the design more safe, consider one of the following:
* Create a `protocol LinearRegression` and implement two separate conforming types: `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression`.
  * `protocol LinearRegression` defines requirement `func fit(data:labels:)` and provides default implementations for `func prediction(for:)` and `func score(data:labels:)` in an extension.
  * `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression` conform to `protocol LinearRegression`, implement `func fit`, and define just the stored properties that they need. This way, `SingleValueDecompositionLinearRegression` won't be able to access `learningRate` or `iterationCount`.
* Keep a single `class LinearRegression`, but define a nested enum representing the "strategy".
  ```swift
  public class LinearRegresion {
      /// A linear regression strategy.
      enum Strategy {
          /// Gradient descent.
          case gradientDescent(learningRate: Float, iterationCount: Int)

          /// Singular value decomposition.
          case singularValueDecomposition
      }
      var strategy: Strategy // replaces `var gradientDescent: Bool`
      ...
  }
  ```
  * Then, create two factory methods: `static func gradientDescent(learningRate: Float, iterationCount: Int) -> LinearRegression` and `static func singleValueDecomposition() -> LinearRegression`. These will be used to construct instances of `LinearRegression`.

cc @rxwei

Reply:
@dan-zheng - I have started working on it.

Reply:
@dan-zheng - I have a quick question for you. What level of documentation should in protocol and implementation?

Reply:
A 1-2 sentence summary per type declaration/function should be good!

For types/functions that implement well-known concepts (e.g. linear regression, matrix pseudoinverse, Minkowski distance), consider adding a reference link (e.g. to Wikipedia).

You can look at [tensorflow/swift-apis](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Loss.swift) as a reference for documentation: see [optimizer](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Optimizers/MomentumBased.swift) documentation.